### PR TITLE
Add `max` aggregation and improve aggregation parsing

### PR DESCRIPTION
**Pull request by ggcampinho**
- Add `Max` aggregation
- Allow field nil for `Terms` aggregation
- Add option `group_max` to `Aggregation.parse`:
  - Allow nested aggregations on parse
  - Allow more than one aggregation

### CODE REVIEW

File: lib/elasticband/aggregation/nested.rb [hunk 1]
```diff
@@ -1,18 +1,24 @@
...
         root_aggregation.to_h.tap do |h|
-          h[root_aggregation.name].merge!(aggs: nested_aggregation.to_h)
+          h[root_aggregation.name].merge!(aggs: nested_hash)
         end
       end
+
+      private
+
+      def nested_hash
+        nested_aggregations.each_with_object({}) { |a, h| h.merge!(a.to_h) }
```
Comment (vnegrisolo) [inline lib/elasticband/aggregation/nested.rb]: this `each` and `merge` is duplicated in: `lib/elasticband/aggregation.rb`. Is there a better way to reuse this code?

File: lib/elasticband/aggregation/terms.rb [hunk 2]
```diff
@@ -5,7 +5,7 @@ class Terms < Aggregation
 
       def initialize(name, field, options = {})
         super(name)
-        self.field = field.to_sym
+        self.field = field.to_sym if field
```
Comment (vnegrisolo) [inline lib/elasticband/aggregation/terms.rb]: why don't you use `try`?

Comment (ggcampinho) [inline lib/elasticband/aggregation/terms.rb]: plain ruby, it's faster and simpler. I prefer `try` when you have multiple things going on

Comment (vnegrisolo) [inline lib/elasticband/aggregation/terms.rb]: ok.. nice way of thinking

File: lib/elasticband/aggregation/nested.rb (hunk 1 above)
Comment (ggcampinho) [inline lib/elasticband/aggregation/nested.rb]: Done in 55a72c8

File: lib/elasticband.rb [hunk 3]
```diff
@@ -1,6 +1,7 @@
 require 'active_support'
 require 'active_support/core_ext/object'
 require 'active_support/core_ext/array'
+require 'active_support/core_ext/hash'
```
Comment (rubensmabueno) [inline lib/elasticband.rb]: You could add in alphabetic order.

Comment (brennovich) [inline lib/elasticband.rb]: Agreed!

File: lib/elasticband.rb [hunk 4]
```diff
@@ -1,6 +1,7 @@
 require 'active_support'
```
Comment (brennovich) [inline lib/elasticband.rb]: Do we really need this?

File: lib/elasticband/aggregation/field_based.rb [hunk 5]
```diff
@@ -0,0 +1,27 @@
+module Elasticband
+  class Aggregation
+    class FieldBased < Aggregation
+      attr_accessor :field, :options
+
+      def initialize(name, field, options = {})
+        super(name)
+        self.field = field.to_sym if field
```
Comment (rubensmabueno) [inline lib/elasticband/aggregation/field_based.rb]: Couldn't we use `self.field &&= field.to_sym`?

File: lib/elasticband/aggregation/max.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+module Elasticband
+  class Aggregation
+    class Max < FieldBased
+      def type
+        :max
```
Comment (rubensmabueno) [inline lib/elasticband/aggregation/max.rb:5]: Can't it be memoized?

File: lib/elasticband/aggregation/nested.rb [hunk 7]
```diff
@@ -1,18 +1,24 @@
 module Elasticband
   class Aggregation
```
Comment (brennovich) [inline lib/elasticband/aggregation/nested.rb:2]: A class that is a namespace, but also a super class which a child class inherits from. Seems complex inheritance stuff :/

File: lib/elasticband.rb (hunk 4 above)
Comment (ggcampinho) [inline lib/elasticband.rb]: Yes, I use array extensions, hash extensions

File: lib/elasticband/aggregation/max.rb (hunk 6 above)
Comment (ggcampinho) [inline lib/elasticband/aggregation/max.rb:5]: Symbols are always the same object. Don't need

File: lib/elasticband.rb (hunk 4 above)
Comment (ggcampinho) [inline lib/elasticband.rb]: Done

File: lib/elasticband.rb (hunk 3 above)
Comment (ggcampinho) [inline lib/elasticband.rb]: Done

File: lib/elasticband/aggregation/max.rb (hunk 6 above)
Comment (brennovich) [inline lib/elasticband/aggregation/max.rb:5]: Not in Ruby 2.2 :)

Symbols are now garbage collected, a simple `freeze` will do the work

Comment (rubensmabueno) [inline lib/elasticband/aggregation/max.rb:5]: http://www.sitepoint.com/symbol-gc-ruby-2-2/

Comment (ggcampinho) [inline lib/elasticband/aggregation/max.rb:5]: > For example :foo and def foo; end both will not be garbage collected, however "foo".to_sym would be eligible for garbage collection.

File: lib/elasticband/aggregation/nested.rb (hunk 7 above)
Comment (ggcampinho) [inline lib/elasticband/aggregation/nested.rb:2]: IMHO as `class` or as `module` we will have the same complexity.

File: lib/elasticband/aggregation/field_based.rb (hunk 5 above)
Comment (ggcampinho) [inline lib/elasticband/aggregation/field_based.rb]: In 3874ea1

Comment (brennovich) [inline lib/elasticband/aggregation/field_based.rb]: :1:

**Comment by ggcampinho** [comment]
@LoveMondays/developers thumbs up?

**Comment by shaneog** [comment]
LGTM :+1:

**Comment by brennovich** [comment]
:+1:

**Pull request merged**